File: src/main/java/de/melsicon/talk/value/simple/pojo/Person.java
package de.melsicon.talk.value.simple.pojo;

import com.google.common.collect.ImmutableSet;
import com.google.errorprone.annotations.Immutable;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;
import java.util.StringJoiner;
import org.checkerframework.checker.nullness.qual.Nullable;

/** Representation of a person. */
@Immutable
public final class Person {
  private final String givenName;

  private final @Nullable String surname;

  private final ImmutableSet<EmailAddress> email;

  /**
   * Creates a Person.
   *
   * @param givenName Set the given name. Required.
   * @param surname Set the surname. Optional, i.e for royalty.
   * @param email Email addresses this person is reachable under. Optional.
   */
  public Person(
      String givenName, @Nullable String surname, @Nullable Iterable<EmailAddress> email) {
    this.givenName = Objects.requireNonNull(givenName, "givenName is a required parameter");
    this.surname = surname;
    this.email = email == null ? ImmutableSet.of() : ImmutableSet.copyOf(email);
  }

  /**
   * Creates a Person.
   *
   * @param givenName Set the given name. Required.
   * @param surname Set the surname. Optional, i.e for royalty.
   * @param email Email addresses this person is reachable under. Optional.
   * @return A person.
   */
  public static Person of(String givenName, @Nullable String surname, String... email) {
    return new Person(
        givenName,
        surname,
        Arrays.stream(email).map(EmailAddress::of).collect(ImmutableSet.toImmutableSet()));
  }

  /**
   * The given name of this person. Everyone has one.
   *
   * @return A name.
   */
  public String givenName() {
    return givenName;
  }

  /**
   * The surname name of this person. Optional, i.e for royalty.
   *
   * @return An optional name
   */
  public Optional<String> surname() {
    return Optional.ofNullable(surname);
  }

  /**
   * Email addresses this person is reachable under.
   *
   * @return An immutable set of email addresses
   */
  public ImmutableSet<EmailAddress> email() {
    return email;
  }

  @Override
  public int hashCode() {
    return Objects.hash(givenName, surname, email);
  }

  @Override
  public boolean equals(@Nullable Object obj) {
    return obj == this
        || (obj instanceof Person p
            && givenName.equals(p.givenName)
            && Objects.equals(surname, p.surname)
            && email.equals(p.email));
  }

  @Override
  public String toString() {
    var emails =
        email.stream()
            .reduce(
                new StringJoiner(", "),
                (joiner, email) -> joiner.add(email.toString()),
                StringJoiner::merge);
    return String.format(
        "Person(givenName=%s surname=%s email=[%s])",
        givenName, surname == null ? "(null)" : surname, emails);
  }
}

File: src/main/scala/de/melsicon/talk/value/simple/scala/Person.scala
package de.melsicon.talk.value.simple.scala

import java.util

import org.checkerframework.checker.nullness.qual.Nullable

import scala.annotation.varargs
import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._

object Person {

  /** Java API: Construct a person
    *
    * @param givenName The given name. Required.
    * @param surname   The surname. Optional, i.e for royalty.
    * @param email     Email addresses this person is reachable under. Optional.
    */
  @varargs
  def of(givenName: String, @Nullable surname: String, email: String*): Person =
    Person(
      util.Objects.requireNonNull(givenName),
      Option(surname),
      email.map(EmailAddress.of)
    )
}

final case class Person(
    givenName: String,
    surname: Option[String] = None,
    email: Seq[EmailAddress] = Seq()
) {

  /** Java API: Construct a person
    *
    * @param givenName The given name. Required.
    * @param surname   The surname. Optional, i.e for royalty.
    * @param email     Email addresses this person is reachable under. Optional.
    */
  def this(
      givenName: String,
      @Nullable surname: String,
      @Nullable email: java.lang.Iterable[EmailAddress]
  ) = {
    this(
      util.Objects.requireNonNull(givenName),
      Option(surname),
      if (email == null) Seq.empty else Seq.from(email.asScala)
    )
  }

  /** Java API: The given name of this person. Everyone has one.
    *
    * @return A name.
    */
  def getGivenName: String = givenName

  /** Java API: The surname of this person. Optional, i.e for royalty.
    *
    * @return An optional name
    */
  def getSurname: util.Optional[String] = surname.toJava

  /** Java API: Email addresses this person is reachable under.
    *
    * @return An immutable list of email addresses
    */
  def getEmail: util.List[EmailAddress] = email.asJava
}
